### PR TITLE
feat: require signatures to prove control of `signer-key` in pox

### LINKED ISSUE #4247: Signer should control which stackers can delegate to the signer
**Is your feature request related to a problem? Please describe.**
For legal or other reasons, signers might not want to accept signing duties for some stackers who want to delegate signing to them.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
@MarvinJanssen mentioned the issue first I think

### PR DESCRIPTION
**Pull request by hstove**
This closes #4247, which is to ensure that all Signers are using signing keys that they either control, or are controlled by the account they're delegating to.

- closes #4247 

This is done by adding a `signer-sig` param to:

- stack-stx
- delegate-stack-stx
- stack-aggregation-commit

The signature follows [SIP018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md) for signed structured data. Following SIP018 has the benefit of built-in wallet support, so signers can use hardware wallets and apps to generate these signatures. 

Following SIP018, the structured data is `{ reward-cycle, reward-cycle }`, where `reward-cycle` is the _current_ stacking cycle where the transaction is made. The `domain` is `{ app: "pox-4-signer", version: "1.0.0", chain-id }`.

Because we'll almost certainly be building the ability for `stacks-signer` to generate these signatures (ie via CLI), I added a SIP18 util library at `/stacksib/src/lib_util/signed_structured_data`. That library has tests that validate the functionality based on the test vectors in SIP018.

I added a function `to_rsv` to `MessageSignature` to match the byte order that Clarity uses for signatures.

Note that I added the new `signer-sig` param as the second-to-last argument because of other conversations around expected argument order, but I'm not really familiar with the requirements there. I also am not sure what other touch points use these functions where the arguments need to be updated. cc @zone117x regarding param order

### CODE REVIEW

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:680]: We use sha512/256 for our message digests

Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:680]: As in sha512(sha256(message))? No problem, I'll change that.

**Comment by hstove** [comment]
Do we need to add the signature check to `delegate-stack-increase` and `delegate-stack-extend`?

**Comment by hstove** [comment]
After further thought, I think we should follow SIP018 (signed structured data) for this. That also unlocks the ability for signers to use a web app to request signatures from a wallet.

**Comment by friedger** [comment]
#4247 is about delegating signer duties, not delegating stx. As far as I understand. 

Pool operators can already refuse to `delegate-stack-stx` for a pool member.

**Comment by hstove** [comment]
@friedger @kantai is it better to describe this ticket as "delegates and delegators need to prove ownership of the signing key"? For delegates, this is proving that the signer they're delegating to has control over the signing key.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:680]: I also notice that in [SIP018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md#formal-specification) (signed structured data) only `sha256` is used, and I'd like to build this off of SIP018.

**Comment by jcnelson** [comment]
_Everyone_ needs to prove ownership of their signing key.  This goes for solo stackers as well, via `stack-stx` and `stack-extend`.  Basically, anywhere you see a signing key as an argument, you need to also require a signature argument and a signature verification (that should be `stack-stx`, `delegate-stack-stx`, `stack-extend`, and `delegate-stack-extend`).

**Comment by friedger** [comment]
Instead of delegate stack stx, pool operators now provide the key in stack aggregation-* function calls.

See #4274

**Comment by hstove** [comment]
> Instead of delegate stack stx, pool operators now provide the key in stack aggregation-* function calls.
> 
> See #4274

Good to know - and I'll have to change this to build off the change from `signer-key` to `signer` too.

**Comment by hstove** [comment]
> _Everyone_ needs to prove ownership of their signing key. This goes for solo stackers as well, via `stack-stx` and `stack-extend`. Basically, anywhere you see a signing key as an argument, you need to also require a signature argument and a signature verification (that should be `stack-stx`, `delegate-stack-stx`, `stack-extend`, and `delegate-stack-extend`).

👍 that's reflected in the current state of the PR

**Comment by hstove** [comment]
There are a number of tests outside of the pox-4 tests related to nakamoto that are failing because of the change to `stack-stx`, I'm working on those still but the pox-4 code is ready for review.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (zone117x) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:574]: Thoughts on removing the `signer-key` arg? Seems redundant now that `signer-sig` is passed and the key can be derived from it? I'm guessing the burn-op versions of these functions will do that to save OP_RETURN space.

Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:574]: Interesting thought. If it weren't for the burn op issue I'd probably push back, because it's too easy for a mis-created signature to cause an invalid signing key to be used. If the message hash differs, then the signer-key will end up as something else.

File: stacks-common/src/util/secp256k1.rs
Comment (jcnelson) [inline stacks-common/src/util/secp256k1.rs:114]: Why is this conversion necessary?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:694]: This should be a constant somewhere

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1274]: Should this be commented?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1650]: Please drop commented code

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:2165]: Can you add some test coverage to this file for all four affected stacking functions to verify that they fail with the expected error code if the signature was bad?  Not just unit-testing the signature verification function; verifying end-to-end that the stacking function fails as expected.  Thanks!

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (jcnelson) [inline stackslib/src/util_lib/signed_structured_data.rs:1]: Missing copyright comment block

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (jcnelson) [inline stackslib/src/util_lib/signed_structured_data.rs:21]: Why does this belong in `util_lib`?  Isn't this only used in testing?  If so, then please remove this file altogether and put this into `stackslib/src/chainstate/stacks/boot/mod.rs`.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: The latest in this PR does follow SIP18, the v1 definitely didn't

File: stacks-common/src/util/secp256k1.rs
Comment (hstove) [reply stacks-common/src/util/secp256k1.rs:114]: Because when recovering or verifying signatures in Clarity, the signature needs to be in RSV order. Whether this helper function is necessary or not is a fine question, it was just helpful to re-use

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (hstove) [reply stackslib/src/util_lib/signed_structured_data.rs:21]: At the moment it's only used in tests, but we'll add it (in a separate PR) to `stacks-signer` so we can expose a CLI function like:

```
stacks-signer create-stacking-signature --stacker S --reward-cycle R
```

I don't have a strong opinion whether it should live here or not, and I'm not the best person to know. But I think it is helpful to expose somewhere that isn't just for tests.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:2592]: Added 4 new tests that cover all the stacking transactions

**Review by setzeus — CHANGES_REQUESTED** [review]
Generally pox_4.clar updates LGTM minus a few syntax issues & one ERR constant inconsistency.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: ah, my habits slipped in here 😬

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: Good call, I also renamed the error message to be more clear (and removed `delegation` as its inaccurate)

**Review by setzeus — APPROVED** [review]

**Comment by hstove** [comment]
FYI I'm going to rebase/build this off of #4269, because that changes the function signatures in pox-4. Shouldn't be too much work and then I'll move the ticket back to review.

**Comment by setzeus** [comment]
@hstove copy, will re-review once re-based.

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (jcnelson) [reply stackslib/src/util_lib/signed_structured_data.rs:37]: Sounds like this belongs in `libsigner` then.  `libsigner` can be a dev-dependency for `stackslib`.

File: stacks-common/src/util/secp256k1.rs
Comment (jcnelson) [reply stacks-common/src/util/secp256k1.rs:114]: Ugh I just noticed that `secp256k1_recover()` and `secp256k_verify()` don't behave the same as the corresponding methods in `Secp256k1PublicKey` and `MessageSignature`.  The former two methods, which Clarity uses, were added by an itinerant 3rd party contributor who couldn't be bothered to figure out how the existing code worked.

**Comment by hstove** [comment]
Ready for review. This was rebased off of #4269 and now off of `next`, which changed a few things from last week's version:

- the `signer-key` (and thus `signer-sig`) args were removed from `delegate-stack-*`
- `signer-sig` was added to `stack-aggregation-commit`
- Now the structured data tuple is `{ pox-addr, reward-cycle }`. Previously it was `{ stacker, reward-cycle }`

File: stacks-common/src/util/secp256k1.rs
Comment (hstove) [reply stacks-common/src/util/secp256k1.rs:114]: Yeah, that's unfortunate because almost every `sign()` implementation returns the signature in RSV, but Clarity uses VRS, so you always have to build a helper method for it. Easy to get tripped up on.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:776]: This method should be removed. It was kept for backwards compatibility. Now, it is not compatible anymore, therefore, we only should use `stack-aggregation-commit-indexed`

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:2330]: bob called delegate-stack-stx that failed here.

Should be changed in line 2323 as well.

**Review by friedger — DISMISSED** [review]
The burn chain tx for stack-stx and stack-extend need to be updated.

Otherwise, just small nits. This looks good if all tests pass.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:776]: I'm curious how many apps use this function vs `-indexed`. My guess is that the easiest path forward is to remove `-indexed` but change the function's return signature here.

For example, looks like the stacks.js package uses `stack-aggregation-commit`.

Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:776]: Stacks.js has not yet been updated to stacks 2.4 🙃

Either is fine to me. Jude might have a stronger opinion.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:2330]: Can you explain more? When I run this test every contract call returns `ok`, including `delegate-stack-stx`

**Comment by hstove** [comment]
> The burn chain tx for stack-stx and stack-extend need to be updated.

We have a ticket for that at #4285, should we wait for that decision to be made before merging this? I'm not sure if we have tests for those burn ops in pox-4. I think it's already broken from just `signer-key` being added, though I might be mistaken.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (zone117x) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:579]: Synthetic pox events are generated using argument indexes rather than named values, so it's generally easier to add new args to the end of the list. Not a deal breaker in this PR because `signer-sig` isn't yet added to the pox events, but I'm working on adding that and our PRs will conflict due to the ordering.

Comment (zone117x) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:579]: Conflict PR: https://github.com/stacks-network/stacks-core/pull/4329

I can modify my PR and change the event arg indexes, no problem. Just an FYI

Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:579]: No problem, let's merge your PR and I'll update mine.

Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:579]: I have no real preference on arg order. Is your PR the only real place we need to worry about arg index? I'd probably just update that synthetic events code to change the signer_key index, but if there are other downstream issues then it's probably better for me to change the argument order on pox-4.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:708]: Similarly, can this take the reward cycle as an argument, and pass it into `get-signer-key-message-hash`?

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:744]: You can use `(current-reward-cycle)` here as the reward cycle arg.

Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:744]: You'd want to pass in the given `reward-cycle` here

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:708]: Also, given the way this is used in `stack-aggregation-commit-indexed`, you'd actually need a caller-supplied reward cycle, since it'll be different than the current reward cycle.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:687]: Given how this is reached from `stack-aggregation-commit-indexed`, you'll want the reward cycle to be an argument anyway -- the reward cycle won't be equal to `current-pox-reward-cycle`.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:776]: It's fine to remove it for pox-4, but please add a changelog entry.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:687]: Actually, I think we're missing some data here.  The signature should really be over the PoX address, the first reward cycle for which the signer will sign, and the number of reward cycles for which the signer will sign.  Without that last parameter, someone could use a signer's key for 12 cycles while the signer might have only wanted to sign for 1 cycle.

* For `stack-stx`, the signature would include `lock-period`
* For `stack-extend`, the signature would include `extend-count`
* For `stack-aggregation-commit-indexed`, the lockup period would just be 1.

**Review by jcnelson — DISMISSED** [review]
I think we want the signature to cover some more data than it does right now.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (zone117x) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:579]: Thanks!

**Review by zone117x — APPROVED** [review]
I'm approving the pox4 function signature and synthetic event changes, thanks @hstove!

At least 2 others should approve the pox4 behavioral changes because I'm not in a position to validate those changes.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: To maintain the behavior of "signer key signatures must use _current_ reward cycle", I've passed in `(- reward-cycle u1)` into the signature validation function

**Comment by hstove** [comment]
I've updated this PR with changes since receiving Jude's feedback. Tagging @jcnelson @setzeus @MarvinJanssen @friedger for reviews.

- The signature message hash now includes two new fields
  - `topic`: one of `stack-stx`, `stack-extend`, or `agg-commit`. This prevents re-using signatures across methods
  - `period`: a `uint` reflecting `lock-period` for stacking or `extend-count` for extending. For `stack-aggregation-commit`, this value is `u1`.
- The `get-signer-key-message-hash` now takes all hash fields as args, instead of dynamically calling `get-current-reward-cycle`. This allows signers to easily call this read-only method to generate a message hash.
- In `stack-aggregation-commit`, the message hash uses the reward cycle provided as user input instead of the current reward cycle.

I've also added some code in `util_lib::signed_structured_data::pox4` to allow re-using the Rust code for generating these signatures instead of re-writing a helper function in different places.

**Review by kantai — COMMENTED** [review]
This LGTM -- I'd just suggest using the `define_named_enum!` macro for the topics, and you should add another unit test that asserts equality between the rust message hash constructor and the contract's read-only fn.

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (hstove) [reply stackslib/src/util_lib/signed_structured_data.rs:100]: :+1: Added in https://github.com/stacks-network/stacks-core/pull/4277/commits/dbacfd06c2c19f91ed531aeac0723cdb02aec248

File: stackslib/src/chainstate/stacks/boot/pox-4.clar
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: Yep good call. I actually just stripped out the entire `let` and have one statement for both methods. The `let` made writing the code easier, but there was no variable re-use so it wasn't necessary. Plus we can save a teeny bit of cost without it!

File: stackslib/src/util_lib/signed_structured_data.rs
Comment (kantai) [inline stackslib/src/util_lib/signed_structured_data.rs:95]: Can `Pox4SignatureTopicOld` be removed?

**Review by kantai — APPROVED** [review]
LGTM -- just see if the `...Old` enum can be removed.

**Comment by hstove** [comment]
> LGTM -- just see if the ...Old enum can be removed.

Ah crap, thanks for catching that. CI should pass and then I think this is good to merge!

**Pull request merged**
